File: serve.js
#!/usr/bin/env node
/**
 * AURA Life Dashboard — servidor local en el puerto 8080 (sin dependencias).
 *
 * Uso:
 *     node serve.js
 *
 * Equivale a: npx http-server -p 8080
 * Las PWAs requieren HTTP (no file://) para Service Worker + manifest.
 */
const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = 8080;
const ROOT = __dirname;

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8',
  '.mjs': 'application/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.webmanifest': 'application/manifest+json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.webp': 'image/webp',
  '.ico': 'image/x-icon',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.txt': 'text/plain; charset=utf-8'
};

const server = http.createServer((req, res) => {
  let urlPath = decodeURIComponent((req.url || '/').split('?')[0]);

  // Seguridad: evitar path traversal
  const safe = path.normalize(urlPath).replace(/^(\.\.[/\\])+/, '');
  let filePath = path.join(ROOT, safe);
  if (!filePath.startsWith(ROOT)) {
    res.writeHead(403);
    return res.end('Forbidden');
  }

  if (urlPath === '/' || urlPath === '') filePath = path.join(ROOT, 'index.html');

  fs.stat(filePath, (err, stat) => {
    if (!err && stat.isDirectory()) filePath = path.join(filePath, 'index.html');
    fs.readFile(filePath, (err2, data) => {
      if (err2) {
        res.writeHead(404, { 'Content-Type': 'text/plain; charset=utf-8' });
        return res.end('404 Not Found: ' + urlPath);
      }
      const ext = path.extname(filePath).toLowerCase();
      res.writeHead(200, {
        'Content-Type': MIME[ext] || 'application/octet-stream',
        'Cache-Control': 'no-cache',
        'Service-Worker-Allowed': '/'
      });
      res.end(data);
    });
  });
});

server.listen(PORT, () => {
  console.log('');
  console.log('  AURA Life Dashboard');
  console.log(`  -> http://localhost:${PORT}`);
  console.log('  Ctrl+C para detener');
  console.log('');
});

File: serve.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
AURA Life Dashboard — servidor local en el puerto 8080.

Uso:
    python serve.py

Equivale a: python -m http.server 8080
Las PWAs requieren HTTP (no file://) para Service Worker + manifest.
Abre http://localhost:8080 automaticamente.
"""
import http.server
import socketserver
import webbrowser
import os
import threading

PORT = 8080
os.chdir(os.path.dirname(os.path.abspath(__file__)))


class Handler(http.server.SimpleHTTPRequestHandler):
    # Sin cache en desarrollo: el SW gestiona el versionado
    def end_headers(self):
        self.send_header("Cache-Control", "no-cache")
        super().end_headers()

    def log_message(self, fmt, *args):
        print("  " + fmt % args)


class ReusableTCPServer(socketserver.TCPServer):
    allow_reuse_address = True


def main():
    url = f"http://localhost:{PORT}"
    with ReusableTCPServer(("0.0.0.0", PORT), Handler) as httpd:
        print()
        print("  AURA Life Dashboard")
        print(f"  -> {url}")
        print("  Ctrl+C para detener")
        print()
        threading.Timer(0.8, lambda: webbrowser.open(url)).start()
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\n  Servidor detenido.")


if __name__ == "__main__":
    main()
